vnext/JSI/Shared/ChakraObjectRef.cpp

```diff
@@ -0,0 +1,325 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#include "ChakraObjectRef.h"
+
+#include "Unicode.h"
+#include "Utilities.h"
+
+#include "jsi//jsi.h"
+
+#include <memory>
+#include <sstream>
+#include <utility>
+
+namespace Microsoft::JSI {
+
+namespace {
+
+inline void CheckedJsAddRef(JsRef ref) {
+  ThrowUponChakraError(JsAddRef(ref, nullptr), "JsAddRef");
+}
+
+inline void CheckedJsRelease(JsRef ref) {
+  ThrowUponChakraError(JsRelease(ref, nullptr), "JsRelease");
+}
+
+} // namespace
+
+void ThrowUponChakraError(JsErrorCode error, const char *const chakraApiName) {
+  if (error != JsNoError) {
+    std::ostringstream errorString;
+    errorString << "A call to "
+                << (chakraApiName ? chakraApiName : "Chakra(Core) API")
+                << " returned error code 0x" << std::hex << error << '.';
+    throw facebook::jsi::JSINativeException(errorString.str());
+  }
+}
+
+ChakraObjectRef::ChakraObjectRef(const ChakraObjectRef &original) noexcept
+    : m_ref{original.m_ref}, m_state{original.m_state} {
+  if (m_state == State::Initialized) {
+    assert(m_ref);
+    CheckedJsAddRef(m_ref);
+  } else {
+    assert(!m_ref);
+  }
+}
+
+ChakraObjectRef::ChakraObjectRef(ChakraObjectRef &&original) noexcept {
+  swap(original);
+}
+
+ChakraObjectRef &ChakraObjectRef::operator=(
+    const ChakraObjectRef &rhs) noexcept {
+  ChakraObjectRef rhsCopy(rhs);
+  swap(rhsCopy);
+  return *this;
+}
+
+ChakraObjectRef &ChakraObjectRef::operator=(ChakraObjectRef &&rhs) noexcept {
+  swap(rhs);
+  return *this;
+}
+
+ChakraObjectRef::~ChakraObjectRef() noexcept {
+  if (m_state == State::Initialized) {
+    assert(m_ref);
+    CheckedJsRelease(m_ref);
+  } else {
+    assert(!m_ref);
+  }
+}
+
+void ChakraObjectRef::Initialize(JsRef ref) {
+  assert(!m_ref);
+
+  if (m_state != State::Uninitialized) {
+    throw facebook::jsi::JSINativeException(
+        "A ChakraObjectRef can only be initialzed once.");
+  }
+
+  if (!ref) {
+    throw facebook::jsi::JSINativeException(
+        "Cannot initialize a ChakraObjectRef with a null reference.");
+  }
+
+  CheckedJsAddRef(ref);
+  m_ref = ref;
+  m_state = State::Initialized;
```

Comment:
>m_state [](start = 2, length = 7)

This is all single threaded, right?

Reply:
Yes. There were plans to make ChakraRuntime thread safe in the future, but I think the switch to V8/Hermes may mean that we no longer have to do this work.

---
In reply to: [336171764](https://github.com/microsoft/react-native-windows/pull/3445#discussion_r336171764) [](ancestors = 336171764)